narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,226 +3,381 @@
+        let mut progression = BTreeMap::new();
+        for (name, _) in self.committee.authorities() {
+            // Last round is 0 (genesis) when authority is not found in store.
+            let last_round = self.certificate_store.last_round_number(name)?.unwrap_or(0);
+            progression.insert(name.clone(), last_round);
+        }
+        Ok(progression)
+    }
+}
+
+/// Fetches certificates from other primaries concurrently, with ~5 sec interval between each request.
+/// Terminates after the 1st successful response is received.
+#[instrument(level = "debug", skip_all)]
+async fn fetch_certificates_helper(
+    name: PublicKey,
+    network: Arc<dyn PrimaryToPrimaryRpc>,
+    committee: Committee,
+    request: FetchCertificatesRequest,
+) -> FetchCertificatesResponse {
+    trace!("Start sending fetch certificates requests");
+    let request_interval = Duration::from_secs(5);
+    let mut peers: Vec<NetworkPublicKey> = committee
+        .others_primaries(&name)
+        .into_iter()
+        .map(|(_, _, network_key)| network_key)
+        .collect();
+    loop {
+        peers.shuffle(&mut ThreadRng::default());
+        let mut fut = FuturesUnordered::new();
+        for peer in peers.iter() {
+            fut.push(network.fetch_certificates(peer, request.clone()));
+            let mut interval = Box::pin(time::sleep(
+                request_interval + Duration::from_millis(ThreadRng::default().gen_range(0..1000)),
+            ));
+            tokio::select! {
+                res = fut.next() => match res {
+                    Some(Ok(resp)) => {
+                        return resp;
+                    }
+                    Some(Err(e)) => {
+                        debug!("Failed to fetch certificates: {e}");
+                    }
+                    None => {}
+                },
+                _ = &mut interval => {
+                    debug!("fetch_certificates_helper: no response within timeout. Sending out a new fetch request.");
+                }
+            };
         }
     }
+}
 
-    fn update_metrics(&self, waiting_len: usize) {
-        self.metrics
-            .pending_elements_certificate_waiter
-            .with_label_values(&[&self.committee.epoch.to_string()])
-            .set(self.pending.len() as i64);
-        self.metrics
-            .waiting_elements_certificate_waiter
-            .with_label_values(&[&self.committee.epoch.to_string()])
-            .set(waiting_len as i64);
+#[instrument(level = "debug", skip_all)]
+async fn store_certificates_helper(
+    response: FetchCertificatesResponse,
+    committee: Committee,
+    certificate_store: &CertificateStore,
+    tx_certificates_loopback: Sender<Certificate>,
+    metrics: Arc<PrimaryMetrics>,
+) -> DagResult<()> {
+    trace!("Start sending fetched certificates to processing");
+    let mut waiters = Vec::new();
+    for certificate in response.certificates {
+        if waiters.len() == MAX_CERTIFICATES_TO_FETCH {
```

Comment:
maybe make this >= to be a little more defensive? I could see a bug being more likely in the future otherwise, if this loop is refactored and no longer pushes only one cert per iteration

Reply:
This code is refactored away now!